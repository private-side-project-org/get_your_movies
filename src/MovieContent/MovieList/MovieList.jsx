import React, { useEffect, useRef } from "react";

import MoviesListItem from "./MovieListItem/MovieListItem";
import SyncLoader from "react-spinners/SyncLoader";
import PropTypes from "prop-types";

import "./movieList.scss";

const propTypes = {
  movies: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
    }).isRequired
  ),
  hasNextPage: PropTypes.bool,
  fetchNextPage: PropTypes.func.isRequired,
  onSetSelectedMovie: PropTypes.func.isRequired,
  selectedTab: PropTypes.string.isRequired,
};

const MovieList = ({
  movies,
  hasNextPage,
  fetchNextPage,
  onSetSelectedMovie,
  selectedTab,
}) => {
  // ref hook to get root node(moviesList-container)
  const ref = useRef();

  // callback to execute pagination
  const callback = (entries) => {
    entries.forEach((entry) => {
      const intersecting = entry.isIntersecting;
      if (intersecting && hasNextPage) {
        fetchNextPage();
      }
    });
  };

  // observer to execute callback when target node(loader) show up 100%(threshold 1) in root node
  const observer = new IntersectionObserver(callback, {
    root: ref.current,
    threshold: 1,
  });

  // set target node(loader) on observer
  useEffect(() => {
    const loader = document.getElementById("loader");
    if (loader) {
      observer.observe(loader);
    }
  }, []);

  return (
    <div ref={ref} className="moviesList-container">
      {movies.map((movie, index) => {
        return (
          <MoviesListItem
            key={movie.id}
            movie={movie}
            isFirstMovie={index === 0}
            onSetSelectedMovie={onSetSelectedMovie}
            selectedTab={selectedTab}
          />
        );
      })}
      {hasNextPage && selectedTab !== "favorite" && (
        <div id="loader" data-test-id="loader">
          <SyncLoader color="#000000" size={15} />
        </div>
      )}
    </div>
  );
};

MovieList.propTypes = propTypes;

MovieList.defaultProps = {
  hasNextPage: false,
};

export default MovieList;
