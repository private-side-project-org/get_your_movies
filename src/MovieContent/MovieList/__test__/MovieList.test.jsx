import React from "react";
import { render } from "@testing-library/react";
import MovieList from "../MovieList";

const movies = [
  {
    id: 1,
    original_title: "test title1",
    overview: "this is test overview1",
    backdrop_path: "abcde12345",
  },
  {
    id: 2,
    original_title: "test title2",
    overview: "this is test overview2",
    backdrop_path: "abcde12345",
  },
  {
    id: 3,
    original_title: "test title3",
    overview: "this is test overview3",
    backdrop_path: "abcde12345",
  },
];

// mocked variables
let hasNextPage = true;
let entries = [{ isIntersecting: true }];
const selectedTab = "search";

// mocked functions
const fetchNextPage = jest.fn();
const onSetSelectedMovie = jest.fn();

// prep mock intersectionObserver
const observe = jest.fn(() => {
  entries.forEach((entry) => {
    const intersecting = entry.isIntersecting;
    if (intersecting && hasNextPage) {
      fetchNextPage();
    }
  });
});

// mock intersectionObserver
window.IntersectionObserver = jest.fn(() => ({
  observe,
}));

let renderedMovieList;

beforeEach(() => {
  renderedMovieList = render(
    <MovieList
      movies={movies}
      hasNextPage={hasNextPage}
      fetchNextPage={fetchNextPage}
      onSetSelectedMovie={onSetSelectedMovie}
      selectedTab={selectedTab}
    />
  );
});

describe("MovieList test", () => {
  it("get props and generate snapshot", () => {
    expect(renderedMovieList.baseElement).toMatchSnapshot();
    hasNextPage = false;
  });

  test("not call fetchnextPage when loader is not on display", () => {
    expect(fetchNextPage).toHaveBeenCalledTimes(1);
    hasNextPage = true;
  });

  test("call fetchnextPage once when loader is on display", () => {
    expect(fetchNextPage).toHaveBeenCalledTimes(2);
  });
});
