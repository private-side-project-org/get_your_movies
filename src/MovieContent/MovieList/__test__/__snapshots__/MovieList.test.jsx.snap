// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList test get props and generate snapshot 1`] = `
<body>
  <div>
    <div
      class="moviesList-container"
    >
      <div
        class="moviesListItem-container"
        data-testid="set-movie-button"
        id="top"
      >
        <h3>
          test title1
        </h3>
        <img
          alt="movie_label"
          src="https://image.tmdb.org/t/p/w300abcde12345"
        />
        <p>
          this is test overview1
        </p>
      </div>
      <div
        class="moviesListItem-container"
        data-testid="set-movie-button"
        id=""
      >
        <h3>
          test title2
        </h3>
        <img
          alt="movie_label"
          src="https://image.tmdb.org/t/p/w300abcde12345"
        />
        <p>
          this is test overview2
        </p>
      </div>
      <div
        class="moviesListItem-container"
        data-testid="set-movie-button"
        id=""
      >
        <h3>
          test title3
        </h3>
        <img
          alt="movie_label"
          src="https://image.tmdb.org/t/p/w300abcde12345"
        />
        <p>
          this is test overview3
        </p>
      </div>
      <div
        data-test-id="loader"
        id="loader"
      >
        <span
          class="css-1xdhyk6"
        >
          <span
            class="css-1j1i60w"
          />
          <span
            class="css-wbrves"
          />
          <span
            class="css-1ctme5z"
          />
        </span>
      </div>
    </div>
  </div>
</body>
`;
